common/http/platform/beast/http.cpp

```diff
@@ -341,5 +379,395 @@ ClientConfig::ClientConfig() {
+		auto err = MakeError(BodyMissingError, "Content-Length is non-zero, but body is missing");
+		CallErrorHandler(err, request_, reply_finished_handler_);
+		return;
+	}
+
+	WriteBody();
+}
+
+void Stream::WriteBody() {
+	auto read = response_->body_reader_->Read(body_buffer_.begin(), body_buffer_.end());
+	if (!read) {
+		CallErrorHandler(read.error(), request_, reply_finished_handler_);
+		return;
+	}
+
+	http_response_->body().data = body_buffer_.data();
+	http_response_->body().size = read.value();
+
+	if (read.value() > 0) {
+		http_response_->body().more = true;
+	} else {
+		// Release ownership of Body reader.
+		response_->body_reader_.reset();
+		http_response_->body().more = false;
+	}
+
+	http::async_write_some(
+		socket_, *http_response_serializer_, [this](const error_code &err, size_t num_written) {
+			WriteBodyHandler(err, num_written);
+		});
+}
+
+void Stream::WriteBodyHandler(const error_code &err, size_t num_written) {
+	if (err) {
+		CallErrorHandler(err, request_, reply_finished_handler_);
+		return;
+	}
+
+	logger_.Trace("Wrote " + to_string(num_written) + " bytes of body data to stream.");
+
+	if (num_written > 0) {
+		// We are still writing the body.
+		WriteBody();
+		return;
+	}
+
+	// We are done.
+	reply_finished_handler_(error::NoError);
+	server_.RemoveStream(shared_from_this());
+}
+
+void Stream::CallBodyHandler() {
+	// Release ownership of writer, which closes it if there are no other holders.
+	request_->body_writer_.reset();
+
+	// Get a pointer to ourselves. This is just in case the body handler make a response, which
+	// it immediately destroys, which would destroy this stream as well. At the end of this
+	// function, it's ok to destroy it.
+	auto stream_ref = shared_from_this();
+
+	server_.body_handler_(request_);
+
+	// MakeResponse() should have been called inside body handler. It can use this to generate a
+	// response, either immediately, or later. Therefore it should still exist, otherwise the
+	// request has not been handled correctly.
+	auto response = maybe_response_.lock();
+	if (!response) {
+		logger_.Error("Handler produced no response. Closing stream prematurely.");
+		server_.RemoveStream(shared_from_this());
+	}
+}
+
+Server::Server(const ServerConfig &server, events::EventLoop &event_loop) :
+	event_loop_(event_loop),
+	acceptor_(GetAsioIoContext(event_loop_)) {
+}
+
+Server::~Server() {
+	Cancel();
+}
+
+error::Error Server::AsyncServeUrl(
+	const string &url, RequestHandler header_handler, RequestHandler body_handler) {
+	auto err = BreakDownUrl(url, address_);
+	if (err) {
+		return MakeError(InvalidUrlError, "Could to server URL " + url + ": " + err.String());
+	}
+
+	boost::system::error_code ec;
+	auto address = asio::ip::make_address(address_.host, ec);
+	if (ec) {
+		return error::Error(
+			ec.default_error_condition(),
+			"Could not construct endpoint from address " + address_.host);
+	}
+
+	asio::ip::tcp::endpoint endpoint(address, address_.port);
+
+	ec.clear();
```

Comment:
I disagree. If something is added in between it would break, but by keeping it close to where I need the cleared error, I know that it will have the correct state.

Reply:
Yup, both approaches have pros and cons. My way of thinking is "use the error or clear it" rather than "let's make sure it's clear before we use it again".